File: backend/app/schemas/voice.py
from typing import Literal

from pydantic import BaseModel, Field

VoiceProviderName = Literal["elevenlabs", "cantoneseai"]


class VoiceTTSRequest(BaseModel):
    text: str = Field(min_length=1)
    language: str = "en"
    preferred_provider: Literal["auto", "elevenlabs", "cantoneseai"] = "auto"
    voice_id: str | None = None


class VoiceTTSResponse(BaseModel):
    request_id: str
    provider: VoiceProviderName
    audio_base64: str
    mime_type: str
    degraded: bool = False
    fallback_reason: str | None = None


class VoiceSTTResponse(BaseModel):
    request_id: str
    provider: VoiceProviderName
    text: str
    degraded: bool = False
    fallback_reason: str | None = None

File: frontend/src/lib/api/voice.ts
import type { TTSProvider } from "@/features/chat/types";
import { apiBaseUrl } from "@/lib/api/base-url";

const API_BASE_URL = apiBaseUrl();

export interface TTSRequest {
  text: string;
  language?: string;
  preferred_provider?: TTSProvider;
}

export interface TTSResponse {
  request_id: string;
  provider: string;
  audio_base64: string;
  mime_type: string;
  degraded: boolean;
  fallback_reason: string | null;
}

export async function postTTS(payload: TTSRequest): Promise<TTSResponse> {
  const response = await fetch(`${API_BASE_URL}/voice/tts`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    cache: "no-store",
    body: JSON.stringify({
      text: payload.text,
      language: payload.language ?? "en",
      preferred_provider: payload.preferred_provider ?? "auto",
    }),
  });

  if (!response.ok) {
    throw new Error(`TTS request failed with status ${response.status}`);
  }

  return (await response.json()) as TTSResponse;
}
